### DISCUSSION TITLE
Pushing to Rust’s tree

### DISCUSSION THREAD

**Discussion by Meziu** [AzureMarker/rust-horizon#3] [General]
Solved the TCP `listen` backlog, we should probably PR the `feature/horizon-std` branch to Rust. I’ve done a check of all the commits we made, and haven’t found anything worth changing. 

@AzureMarker @ian-h-chamberlain if you have a bit of time you should do this too, so we can make sure everything’s good.

**Comment by ian-h-chamberlain** [comment]
Took a look through the diff, I found a couple minor cleanup things I will open a PR for shortly.
Edit: #4 opened

One thing I noticed: `sys::env::os::EXE_EXTENSION` appears to be empty by default on most other platforms (including Linux), but is set to `"elf"` for Horizon. Should we use `""` to match other platforms, since as far as I know it's not strictly required to have `.elf` extension?

Edit: one other thing, I ran `./x.py test library/std` for my host (macOS) just as a double check that we didn't break anything, has anyone else had a chance to try this for a different host? CI will cover it but we might save ourselves some pain if we catch things beforehand.

**Reply by Meziu** [reply]
> `sys::env::os::EXE_EXTENSION` appears to be empty by default on most other platforms (including Linux), but is set to `"elf"` for Horizon. Should we use `""` to match other platforms, since as far as I know it's not strictly required to have `.elf` extension?

No, we should keep the elf type. It is standard in our toolchain (as well as many other ARM targets) to have the `elf` suffix. This is how `devkitPRO` has handled it too. Any other value would be misleading to the user, in most cases at least.

Also, yeah, I should run the tests too.

**Reply by AzureMarker** [reply]
I think the odd thing is that we set it but don't set `EXE_SUFFIX`. For the platforms that do set `EXE_EXTENSION`, they also set `EXE_SUFFIX`.

**Reply by AzureMarker** [reply]
All tests passed on my machine (linux x86_64).

**Reply by AzureMarker** [reply]
I pushed on more small cleanup: https://github.com/AzureMarker/rust-horizon/commit/ed0f5767aabbb8573141d4a9c706268f04651b15

Once this EXE thing is resolved I can make a PR to upstream.

**Reply by AzureMarker** [reply]
I'll also rebase the branches on latest rust-lang master.

Edit: branches have been rebased.

**Reply by Meziu** [reply]
> I think the odd thing is that we set it but don't set `EXE_SUFFIX`. For the platforms that do set `EXE_EXTENSION`, they also set `EXE_SUFFIX`.

Well, the suffix has nothing to do with the extension, does it? No platform I know of actually uses suffixes for executables, but it is common for libraries.

**Reply by ian-h-chamberlain** [reply]
> Well, the suffix has nothing to do with the extension, does it? No platform I know of actually uses suffixes for executables, but it is common for libraries.

I'm not sure that is the kind of suffix meant by `EXE_SUFFIX`... From a quick search, Windows, SGX, WASM, and Emscriptem all use both suffix and extension and they seem be the same as each other with the exception of a dot in the suffix. 

Edit: [the docs](https://doc.rust-lang.org/std/env/consts/constant.EXE_SUFFIX.html) are not very clear on the difference between `_SUFFIX` and `_EXTENSION` either; for `DLL_SUFFIX` the example values match exactly `DLL_EXTENSIONS`, but for `EXE_SUFFIX` they provide some examples (`.nexe`, `.pexe`) that are not found in the `EXE_EXTENSION` docs, and I'm not familiar with these particular extensions...

**Reply by Meziu** [reply]
I believed it was:
 _prefix_ + **name** + _suffix_ + ._extension_

I honestly don’t have any idea of why anyone would want both suffix and extension other than for this format, but maybe we are missing something…

**Reply by ian-h-chamberlain** [reply]
Lol, I agree that would make more sense, but perhaps it's some historical reason. I suppose Rust maintainers would tell us if we opened a PR that didn't match their expectations.

**Reply by ian-h-chamberlain** [reply]
Asked about it [on Zulip](https://rust-lang.zulipchat.com/#narrow/stream/122651-general/topic/std.3A.3Aenv.3A.3Aconsts.3A.3A*_EXTENSION.20vs.20*_SUFFIX), seems like keeping them both the same is probably what we want.

**Reply by Meziu** [reply]
Opened #5 to fix this issue based on your findings.

**Reply by Meziu** [reply]
> Once this EXE thing is resolved I can make a PR to upstream.

Well, I think it is finally time 😁

**Reply by AzureMarker** [reply]
I should be able to open a PR this weekend.

**Reply by AzureMarker** [reply]
Opened! https://github.com/rust-lang/rust/pull/95897

**Reply by Meziu** [reply]
@AzureMarker we should probably assign a different reviewer on https://github.com/rust-lang/rust/pull/95897, it looks like the current one won’t reply…
